tests/MongoDB.Driver.Core.Tests/Core/Operations/AsyncCursorTests.cs

```diff
@@ -315,6 +498,125 @@ private ConnectionDescription CreateConnectionDescriptionSupportingSession()
                 maxTime.WithDefault(null));
         }
+
+        private void SetupChannelMocks(Mock<IChannelSource> mockChannelSource, Mock<IChannelHandle> mockChannelHandle, bool async, string commandResult, string version = "3.6.0")
+        {
+            mockChannelHandle
+                .Setup(c => c.ConnectionDescription)
+                .Returns(CreateConnectionDescriptionSupportingSession(version));
+
+            var bsonCommandResult = BsonDocument.Parse(commandResult);
+            if (async)
+            {
+                mockChannelSource
+                    .Setup(c => c.GetChannelAsync(It.IsAny<CancellationToken>()))
+                    .ReturnsAsync(mockChannelHandle.Object);
+
+                mockChannelHandle
+                    .Setup(
+                        c => c.CommandAsync(
+                            It.IsAny<ICoreSession>(),
+                            It.IsAny<ReadPreference>(),
+                            It.IsAny<DatabaseNamespace>(),
+                            It.IsAny<BsonDocument>(),
+                            It.IsAny<IEnumerable<Type1CommandMessageSection>>(),
+                            It.IsAny<IElementNameValidator>(),
+                            It.IsAny<BsonDocument>(),
+                            It.IsAny<Action<IMessageEncoderPostProcessor>>(),
+                            It.IsAny<CommandResponseHandling>(),
+                            It.IsAny<IBsonSerializer<BsonDocument>>(),
+                            It.IsAny<MessageEncoderSettings>(),
+                            It.IsAny<CancellationToken>()))
+                    .ReturnsAsync(bsonCommandResult);
+            }
+            else
+            {
+                mockChannelSource
+                    .Setup(c => c.GetChannel(It.IsAny<CancellationToken>()))
+                    .Returns(mockChannelHandle.Object);
+
+                mockChannelHandle
+                    .Setup(
+                        c => c.Command(
+                            It.IsAny<ICoreSession>(),
+                            It.IsAny<ReadPreference>(),
+                            It.IsAny<DatabaseNamespace>(),
+                            It.IsAny<BsonDocument>(),
+                            It.IsAny<IEnumerable<Type1CommandMessageSection>>(),
+                            It.IsAny<IElementNameValidator>(),
+                            It.IsAny<BsonDocument>(),
+                            It.IsAny<Action<IMessageEncoderPostProcessor>>(),
+                            It.IsAny<CommandResponseHandling>(),
+                            It.IsAny<IBsonSerializer<BsonDocument>>(),
+                            It.IsAny<MessageEncoderSettings>(),
+                            It.IsAny<CancellationToken>()))
+                    .Returns(bsonCommandResult);
+            }
+        }
+
+        private void VerifyHowManyTimesKillCursorWasCalled(Mock<IChannelHandle> mockChannelHandle, Times times, bool async, bool checkKillCommand = true)
```

Comment:
Cool! I like this approach as it's not tied closely to the implementation. Nice! 😸 
One minor comment: perhaps it might be worth  refactoring this into two methods? e.g. `VerifyHowManyTimesKillCursorWasCalled` and `VerifyHowManyTimesCommandWasCalled`  Currently the boolean "checkKillCommand" completely changes what the method checks.

Also, it might be worth it to check that the command is "killCursors" when checking that the command was executed.

Overall, I really like this :)